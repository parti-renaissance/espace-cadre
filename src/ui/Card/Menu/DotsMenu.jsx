import React, { useState } from 'react'
import { IconButton, Menu, MenuItem as MuiMenuItem } from '@mui/material'
import MoreVertIcon from '@mui/icons-material/MoreVert'
import { styled } from '@mui/system'
import PropTypes from 'prop-types'
import Loader from '~/ui/Loader'

const Wrapper = styled('div')`
  display: flex;
`

const MenuItem = styled(MuiMenuItem)(
  ({ theme }) => `
    font-size: 13px;
    padding: ${theme.spacing(0.5, 1)};
    background-color: ${theme.palette.whiteCorner};
    &:hover {
      background-color: ${theme.palette.gray100};
    },
  `
)

export const DotsMenuItem = ({ onClick, closeMenu, children, cancelLoader = false, deleteLoader = false }) => {
  const handleClick = async () => {
    await onClick()
    closeMenu()
  }

  return (
    <MenuItem onClick={handleClick}>
      {cancelLoader && (
        <>
          <Loader size={12} />
          &nbsp;
        </>
      )}
      {deleteLoader && (
        <>
          <Loader size={12} />
          &nbsp;
        </>
      )}
      {children}
    </MenuItem>
  )
}

DotsMenuItem.propTypes = {
  onClick: PropTypes.func.isRequired,
  closeMenu: PropTypes.func,
  children: PropTypes.node.isRequired,
  cancelLoader: PropTypes.bool,
  deleteLoader: PropTypes.bool,
}

const DotsMenu = ({ children }) => {
  const [anchorEl, setAnchorEl] = useState(null)

  const handleClose = () => {
    setAnchorEl(null)
  }

  const handleClick = e => {
    setAnchorEl(e.currentTarget)
  }

  return (
    <Wrapper data-cy="dot-action-menu">
      <IconButton size="small" onClick={handleClick} sx={{ p: 0 }}>
        <MoreVertIcon />
      </IconButton>
      <Menu
        transitionDuration={0}
        anchorEl={anchorEl}
        keepMounted
        open={!!anchorEl}
        onClose={handleClose}
        sx={{
          fontSize: '13px',
          padding: 0,
          borderRadius: '8px',
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'center',
        }}
      >
        {React.Children.map(children, c => c && React.cloneElement(c, { closeMenu: handleClose }))}
      </Menu>
    </Wrapper>
  )
}

DotsMenu.propTypes = {
  children: PropTypes.node.isRequired,
}

export default DotsMenu
