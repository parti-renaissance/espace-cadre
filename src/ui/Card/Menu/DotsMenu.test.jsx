import { fireEvent, render, screen } from '@testing-library/react'
import DotsMenu, { DotsMenuItem } from 'ui/Card/Menu/DotsMenu'

vi.mock('@mui/system', () => ({
  styled: c => () => c,
}))
vi.mock('@mui/icons-material/MoreVert', () => ({ default: () => <div className="mui-MoreVertIcon-mock" /> }))
vi.mock('@mui/material', () => ({
  Button: ({ children }) => <div className="mui-button-mock">{children}</div>,
  IconButton: ({ children }) => <div className="mui-button-mock">{children}</div>,
  Menu: ({ children }) => <div className="mui-Menu-mock">{children}</div>,
  MenuItem: ({ children, onClick }) => (
    <div className="mui-MenuItem-mock" onClick={onClick} data-testid="mui-MenuItem-mock">
      {children}
    </div>
  ),
}))
describe('DotsMenu', () => {
  it('displays a DotsMenu', () => {
    const mockOnClick = vi.fn()
    const { container } = render(
      <DotsMenu>
        <DotsMenuItem onClick={mockOnClick}>foo</DotsMenuItem>
      </DotsMenu>
    )
    expect(container).toMatchSnapshot()
  })
  it('calls a mock when clicking on menu item', () => {
    const mockOnClick = vi.fn()
    render(
      <DotsMenu>
        <DotsMenuItem onClick={mockOnClick}>foo</DotsMenuItem>
      </DotsMenu>
    )

    const menuItem = screen.getByTestId('mui-MenuItem-mock')
    fireEvent.click(menuItem)
    expect(mockOnClick).toHaveBeenCalled()
  })
})
