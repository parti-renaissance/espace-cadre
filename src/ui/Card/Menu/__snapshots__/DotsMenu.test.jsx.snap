// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DotsMenu > displays a DotsMenu 1`] = `
<div>
  <div
    data-cy="dot-action-menu"
  >
    <div
      class="mui-button-mock"
    >
      <div
        class="mui-MoreVertIcon-mock"
      />
    </div>
    <div
      class="mui-Menu-mock"
    >
      <div
        class="mui-MenuItem-mock"
        data-testid="mui-MenuItem-mock"
      >
        foo
      </div>
    </div>
  </div>
</div>
`;
